File: python/med/Question3.py
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = left = 0
        # 哈希表写法
        # cnt = defaultdict(int)
        # for right, ch in enumerate(s):
        #     cnt[ch] += 1
        #     while cnt[ch] > 1:
        #         cnt[s[left]] -= 1
        #         left += 1

        # 哈希集合写法
        window = set()
        for right, ch in enumerate(s):
            # 要先删再加，否则无法识别出重复
            while ch in window:
                window.remove(s[left])
                left += 1
            window.add(ch)
            ans = max(ans, right - left + 1)
        return ans


print(Solution().lengthOfLongestSubstring("abcabcbb"))

File: cpp/med/Question3.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;
        int max_len = 0, left = 0;
        for (int right = 0; right < s.length(); right++) {
            while (set.find(s[right]) != set.end()) {
                set.erase(s[left]);
                left++;
            }
            set.insert(s[right]);
            max_len = max(max_len, right - left + 1);
        }
        return max_len;
    }
};

int main() {
    string s = "abcabcbb";
    cout << Solution().lengthOfLongestSubstring(s) << endl;
    return 0;
}
